src/seq/mod.rs

```diff
@@ -191,17 +191,7 @@ pub trait IteratorRandom: Iterator + Sized {
         let mut result = None;
 
         if upper == Some(lower) {
-            // Remove this once we can specialize on ExactSizeIterator
             return if lower == 0 { None } else { self.nth(rng.gen_range(0, lower)) };
-        } else if lower <= 1 {
-            result = self.next();
-            if result.is_none() {
-                return result;
-            }
-            consumed = 1;
-            let hint = self.size_hint();
-            lower = hint.0;
-            upper = hint.1;
```

Comment:
I don't have a strong opinion on this one. I answered over in #593. In theory this change will slow down unhinted iterators, but speed up iterators with "unusual" hinting (such as the "window hinted" one in the tests). But in both cases the change should be quite small. And of course less code is always a win.

Reply:
It's difficult tying much to theory when it comes to optimisations isn't it? Perhaps simpler code works better with the branch predictor. It's actually the unhinted one that's faster.

Reply:
Cool. I agree that using too much theory for optimizations can be dangerous. But this is likewise true for using microbenchmarks.

Either way, the intended win from this code is really small, so I'm totally ok with it being removed.